internal/arn/arn.go

```diff
@@ -0,0 +1,146 @@
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package arn
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+)
+
+// ARN structure:
+//
+// arn:partition:service:region:account-id:resource-type/resource-id
+//
+// In this implementation, account-id is empty.
+//
+// Reference: https://docs.aws.amazon.com/general/latest/gr/aws-arns-and-namespaces.html
+
+type arnPartition string
+
+const (
+	arnPartitionMinio arnPartition = "minio"
+)
+
+type arnService string
+
+const (
+	arnServiceIAM arnService = "iam"
+)
+
+type arnRegion string
+
+const (
+	arnRegionDefault arnRegion = ""
+)
+
+type arnResourceType string
+
+const (
+	arnResourceTypeRole arnResourceType = "role"
+)
+
+// ARN - representation of resources based on AWS ARNs.
+type ARN struct {
+	Partition    arnPartition
+	Service      arnService
+	Region       arnRegion
+	ResourceType arnResourceType
+	ResourceID   string
+}
+
+var (
+	// Allows lower-case chars, numbers, '.', '-', '_' and '/'. Starts with a letter or digit.
+	validResourceIDRegex = regexp.MustCompile(`^[a-z0-9][a-z0-9_/\.-]*$`)
+)
+
+// NewIAMRoleARN - returns an ARN for a role in MinIO.
+func NewIAMRoleARN(resourceID string) (ARN, error) {
+	if !validResourceIDRegex.MatchString(resourceID) {
+		return ARN{}, fmt.Errorf("Invalid resource ID: %s", resourceID)
+	}
+	return ARN{
+		Partition:    arnPartitionMinio,
+		Service:      arnServiceIAM,
+		Region:       arnRegionDefault,
+		ResourceType: arnResourceTypeRole,
+		ResourceID:   resourceID,
+	}, nil
+}
+
+// String - returns string representation of the ARN.
+func (arn ARN) String() string {
+	return strings.Join(
+		[]string{
+			"arn",
+			string(arn.Partition),
+			string(arn.Service),
+			string(arn.Region),
+			"", // account-id is always empty in this implementation
+			string(arn.ResourceType) + "/" + arn.ResourceID,
+		},
+		":",
+	)
+}
+
+// Parse - parses an ARN string into a type.
+func Parse(s string) (arn ARN, err error) {
+	ps := strings.Split(s, ":")
+	if len(ps) != 6 ||
+		ps[0] != "arn" {
+		err = fmt.Errorf("Invalid ARN string format")
+		return
+	}
+
+	if ps[1] != string(arnPartitionMinio) {
+		err = fmt.Errorf("Invalid ARN - bad partition field")
+		return
+	}
+
+	if ps[2] != string(arnServiceIAM) {
+		err = fmt.Errorf("Invalid ARN - bad service field")
+		return
+	}
+
+	// ps[3] is region - not validated.
```

Comment:
You can validate the region if the server has a region configured @donatello - you would have to pass the region as an input param here.

Reply:
I've updated it so that the server generates ARNs with the configured region (value of `globalServerRegion`). We still dont need to validate it here, it just needs to be parsed and sent out. It will be validated when the server checks if the provided ARN in STS API is present in server configuration.